Objects/dictobject.c

```diff
@@ -7333,45 +7332,68 @@ set_or_clear_managed_dict(PyObject *obj, PyObject *new_dict, bool clear)
                             (PyDictObject *)Py_XNewRef(new_dict));
 
         Py_END_CRITICAL_SECTION();
-        decref_maybe_delay((PyObject *)dict, !clear);
+        decref_maybe_delay((PyObject *)dict, true);
     }
     assert(_PyObject_InlineValuesConsistencyCheck(obj));
     return err;
 }
 
-int
-_PyObject_SetManagedDict(PyObject *obj, PyObject *new_dict)
-{
-    return set_or_clear_managed_dict(obj, new_dict, false);
-}
-
 void
 PyObject_ClearManagedDict(PyObject *obj)
 {
-    if (set_or_clear_managed_dict(obj, NULL, true) < 0) {
-        /* Must be out of memory */
-        assert(PyErr_Occurred() == PyExc_MemoryError);
-        PyErr_FormatUnraisable("Exception ignored while "
-                               "clearing an object managed dict");
-        /* Clear the dict */
+    // This is called when the object is being freed and therefore
+    // has no other references or during GC when the world is stopped
+    // so we don't need any locking.
```

Comment:
I think it would be worth adding an assert to check this condition.

Reply:
Unfortunately I don't think there's a strong assert we can make here. As @colesbury points out, the world isn't stopped, and because the object could be part of cyclical garbage it can basically have an arbitrary reference count :(